pipeline/covfefe.py

```diff
@@ -0,0 +1,121 @@
+import argparse
+from soopercool import BBmeta
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+
+def covFeFe(args):
+    """
+    This function handles the covariance matrix computation.
+    We can use a set of simulations to build a numerical estimate
+    of the covariance matrices.
+    TODO: implement an analytical estimate of the covariances
+    """
+    meta = BBmeta(args.globals)
+    cov_dir = meta.covmat_directory
+    n_bins = meta.get_n_bandpowers()
+
+    Nsims = meta.num_sims
+    cl_dir = meta.cell_sims_directory
+
+    field_pairs = [f"{m1}{m2}" for m1 in "TEB" for m2 in "TEB"]
+
+    cross_ps_names = meta.get_ps_names_list(type="all", coadd=True)
+
+    # Build the covariance matrix elements to compute
+    cov_names = []
+    for i, (ms1, ms2) in enumerate(cross_ps_names):
+        for j, (ms3, ms4) in enumerate(cross_ps_names):
+            if i > j:
+                continue
+            cov_names.append((ms1, ms2, ms3, ms4))
+
+    # Load the simulations
+    cl_dict = {}
+    for ms1, ms2 in cross_ps_names:
+        cl_dict[ms1, ms2] = []
+        for iii in range(Nsims):
+            cells_dict = np.load(
+                f"{cl_dir}/decoupled_cross_pcls_nobeam_{ms1}_{ms2}_{iii:04d}.npz", # noqa
+            )
+            cl_vec = np.concatenate(
+                [
+                    cells_dict[field_pair] for field_pair in field_pairs
+                ]
+            )
+            cl_dict[ms1, ms2].append(cl_vec)
+
+        cl_dict[ms1, ms2] = np.array(cl_dict[ms1, ms2])
+
+    full_cov_dict = {}
+    for ms1, ms2, ms3, ms4 in cov_names:
+
+        cl12 = cl_dict[ms1, ms2]
+        cl34 = cl_dict[ms3, ms4]
+
+        cl12_mean = np.mean(cl12, axis=0)
+        cl34_mean = np.mean(cl34, axis=0)
+        cov = np.mean(
+            np.einsum("ij,ik->ijk", cl12-cl12_mean, cl34-cl34_mean),
+            axis=0
+        )
+        full_cov_dict[ms1, ms2, ms3, ms4] = cov
+
```

Comment:
At this point, it could be advantageous to add the option of truncating the sub- and superdiagonals larger than order k in every [n_bins, n_bins] covariance block similarly to what was done [here](https://github.com/simonsobs/BBPower/blob/461cd6e8bc187a7a3d6633f2ccac64a27fdcd8a7/bbpower/power_summarizer.py#L31). This would allow a study of maximizing the data compression by saving a maximally sparse covariance matrix (without losing information).

Reply:
Thanks for the comment. We might want to have this option in the `sacc` file creation (next step to be implemented in the pipeline) to avoid re-running covariance estimation every time.